datafusion/physical-plan/src/memory.rs

```diff
@@ -144,6 +144,9 @@ pub trait LazyBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
 
     /// Generate the next batch, return `None` when no more batches are available
     fn generate_next_batch(&mut self) -> Result<Option<RecordBatch>>;
+
+    /// Returns a new instance with the state reset.
+    fn reset_state(&self) -> Arc<RwLock<dyn LazyBatchGenerator>>;
```

Comment:
I don't know if there is a better way to define this method. Returning a `Box<>` might be better, but then I don't think it's possible to move it to `Arc<RwLock<>>`.